api/src/main/java/org/apache/iceberg/metrics/FixedReservoirHistogram.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.metrics;
+
+import java.util.Arrays;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/**
+ * A {@link Histogram} implementation that uses a fixed size reservoir like circular buffer. Oldest
+ * observation is replaced by latest observation.
+ */
+public class FixedReservoirHistogram implements Histogram {
+  private final long[] measurements;
+  private long count;
+
+  public FixedReservoirHistogram(int reservoirSize) {
+    this.measurements = new long[reservoirSize];
+    this.count = 0L;
+  }
+
+  @Override
+  public synchronized long count() {
+    return count;
+  }
+
+  @Override
+  public synchronized void update(long value) {
```

Comment:
I think this should be:

```java
  public synchronized void update(long value) {
    count += 1;
    index = count < measurements.length ? count : rand.nextInt(count);
    if (index < measurements.length) {
      measurements[index] = value;
    }
  }
```

Reply:
as mentioned in another comment, we may need to provide two ways of tracking observations
* circular buffer mode with bias toward recent data, which is good for streaming jobs
* uniform sampling with like the code snippet above, which probably works better for batch jobs